### PR TITLE
ISPN-6069 Unify the continuous query API for remote and embedded mode

### PR DESCRIPTION
**Pull request by anistor**
Jira: https://issues.jboss.org/browse/ISPN-6069

### CODE REVIEW

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/event/ContinuousQueryImpl.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+   public void removeContinuousQueryListener(ContinuousQueryListener<K, ?> listener) {
+      for (Iterator<ClientEntryListener> it = listeners.iterator(); it.hasNext(); ) {
+         ClientEntryListener l = it.next();
+         if (l.listener == listener) {
+            cache.removeClientListener(l);
+            it.remove();
+            break;
+         }
+      }
+   }
```
Comment (gustavonalle) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/event/ContinuousQueryImpl.java]: Does this method need to be java 6 compliant?

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/event/ContinuousQueryImpl.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+   public void removeAllListeners() {
+      for (ClientEntryListener l : listeners) {
+         cache.removeClientListener(l);
+      }
+      listeners.clear();
+   }
+
+   @ClientListener(filterFactoryName = Filters.CONTINUOUS_QUERY_FILTER_FACTORY_NAME,
+         converterFactoryName = Filters.CONTINUOUS_QUERY_FILTER_FACTORY_NAME,
+         useRawData = true, includeCurrentState = true)
```
Comment (gustavonalle) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/event/ContinuousQueryImpl.java]: Not really strictly related to this PR, but WDYT about deprecating the ```KeyValueFilter``` and the ```Converter``` from ```@ClientListener``` and from continuous query, replacing both by a single ```KeyValueFilterConverter```?

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/event/ContinuousQueryImpl.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+      public void handleClientCacheEntryCreatedEvent(ClientCacheEntryCustomEvent<byte[]> event) throws IOException {
+         ContinuousQueryResult cqr = ProtobufUtil.fromByteArray(serializationContext, event.getEventData(), ContinuousQueryResult.class);
+         Object key = ProtobufUtil.fromWrappedByteArray(serializationContext, cqr.getKey());
+         Object value = cqr.getValue() != null ? ProtobufUtil.fromWrappedByteArray(serializationContext, cqr.getValue()) : cqr.getProjection();
+         if (cqr.isJoining()) {
+            listener.resultJoining(key, value);
+         } else {
+            listener.resultLeaving(key);
+         }
+      }
```
Comment (gustavonalle) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/event/ContinuousQueryImpl.java]: Nitpicking:
``` return cqr.isJoining() ? listener.resultJoining(key, value) : listener.resultLeaving(key); ```

File: query-dsl/src/main/java/org/infinispan/query/dsl/continuous/ContinuousQuery.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+public interface ContinuousQuery<K, V> {
+
+   <C> void addContinuousQueryListener(Query query, ContinuousQueryListener<K, C> listener);
+
+   void removeContinuousQueryListener(ContinuousQueryListener<K, ?> listener);
+
+   List<ContinuousQueryListener<K, ?>> getListeners();
+
+   void removeAllListeners();
+}
```
Comment (gustavonalle) [inline query-dsl/src/main/java/org/infinispan/query/dsl/continuous/ContinuousQuery.java]: Are ```ContinuousQuery``` and ```ContinuousQueryListener``` really part of the query dsl ? Probably they should live on ```org.infinispan.query.continuous```, or maybe the artifact ```query-dsl``` should be called ```query-commons```, wdyt?

**Comment by gustavonalle** [comment]
@anistor nice refactor :+1:

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/event/ContinuousQueryImpl.java (hunk 1 above)
Comment (anistor) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/event/ContinuousQueryImpl.java]: No.

File: query-dsl/src/main/java/org/infinispan/query/dsl/continuous/ContinuousQuery.java (hunk 4 above)
Comment (anistor) [inline query-dsl/src/main/java/org/infinispan/query/dsl/continuous/ContinuousQuery.java]: They are not really part of the DSL. I'll move them to ```org.infinispan.query.continuous``` package. I think we could rename ```query-dsl``` module to ```query-api``` or something like that.

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/event/ContinuousQueryImpl.java (hunk 2 above)
Comment (anistor) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/event/ContinuousQueryImpl.java]: I think we can solve this easily in a backward compatible manner. If ```filterFactoryName``` is specified and the factory produces CacheEventFilterConverterFactory then we should internally do both filtering and conversion (in same step) even if no ```converterFactoryName``` is specified. Same if both attributes are specified and are equal. If both attributes are specified and are not equal then we should do separate filtering and conversion. 

This is how core already handles filtering + conversion in same step. Not sure why client events are handled a bit differently.

File: query-dsl/src/main/java/org/infinispan/query/dsl/continuous/ContinuousQuery.java (hunk 4 above)
Comment (anistor) [inline query-dsl/src/main/java/org/infinispan/query/dsl/continuous/ContinuousQuery.java]: I remember why I did not initally move them to ```org.infinispan.query.continuous```. OSGI hates split packages. So I'm a bit out of ideas.

Comment (gustavonalle) [inline query-dsl/src/main/java/org/infinispan/query/dsl/continuous/ContinuousQuery.java]: Maybe ```org.infinispan.query.api.continuous``` ?

Comment (anistor) [inline query-dsl/src/main/java/org/infinispan/query/dsl/continuous/ContinuousQuery.java]: I moved the CQ interfaces to ```org.infinispan.query.api.continuous``` package in query-dsl module. This required some hacking of the osgi imports/exports. 

I'd also rename query-dsl module to query-api in future.

**Comment by gustavonalle** [comment]
looks good @anistor, integrating...

**Comment by gustavonalle** [comment]
integrated, thanks @anistor

**Pull request closed without merging**